File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/core/PrivacySettings.kt
// C12 — Kotlin mirror of FlagshipCore/PrivacySettings.swift.
//
// Persisted user preferences governing how the app gates access to
// its content. Today the only toggle is biometric-at-launch.

package com.flagshipserver.app.core

import android.content.Context
import android.content.SharedPreferences
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

private const val PRIVACY_PREFS = "flagship.privacy"
private const val KEY_REQUIRE_BIOMETRIC = "requireBiometricAtLaunch"
private const val KEY_REQUIRE_PASSPHRASE = "requirePassphraseAtLaunch"
private const val KEY_THEME_MODE = "themeMode"

/** App appearance: follow the system, or force light / dark. */
enum class ThemeMode { AUTO, LIGHT, DARK }

class PrivacySettings(private val prefs: SharedPreferences) {

    private val _requireBiometricAtLaunch =
        MutableStateFlow(prefs.getBoolean(KEY_REQUIRE_BIOMETRIC, true))

    /** True iff a BiometricPrompt evaluation is required each time the
     *  app cold-launches or returns from background. Defaults ON — a
     *  restored account opens behind a biometric unlock rather than a
     *  full sign-in; the user can turn it off to open straight in.
     *  `getBoolean` returns the default only when the key is unset, so an
     *  explicit choice is always honoured. */
    val requireBiometricAtLaunch: StateFlow<Boolean> =
        _requireBiometricAtLaunch.asStateFlow()

    fun setRequireBiometricAtLaunch(value: Boolean) {
        prefs.edit().putBoolean(KEY_REQUIRE_BIOMETRIC, value).apply()
        _requireBiometricAtLaunch.value = value
    }

    private val _requirePassphraseAtLaunch =
        MutableStateFlow(prefs.getBoolean(KEY_REQUIRE_PASSPHRASE, false))

    /** The strictest option: when ON, the app does NOT restore the
     *  persisted session on launch, so every open requires a full
     *  sign-in (the account passphrase), not just a biometric unlock.
     *  Default OFF. Supersedes [requireBiometricAtLaunch] when both set. */
    val requirePassphraseAtLaunch: StateFlow<Boolean> =
        _requirePassphraseAtLaunch.asStateFlow()

    fun setRequirePassphraseAtLaunch(value: Boolean) {
        prefs.edit().putBoolean(KEY_REQUIRE_PASSPHRASE, value).apply()
        _requirePassphraseAtLaunch.value = value
    }

    private val _themeMode = MutableStateFlow(
        runCatching { ThemeMode.valueOf(prefs.getString(KEY_THEME_MODE, null) ?: "AUTO") }
            .getOrDefault(ThemeMode.AUTO),
    )

    /** Chosen app appearance. Default AUTO (follow the system). The theme
     *  wrapper in MainActivity reads this to pick the light/dark palette. */
    val themeMode: StateFlow<ThemeMode> = _themeMode.asStateFlow()

    fun setThemeMode(value: ThemeMode) {
        prefs.edit().putString(KEY_THEME_MODE, value.name).apply()
        _themeMode.value = value
    }

    companion object {
        fun fromContext(ctx: Context): PrivacySettings {
            val prefs = ctx.getSharedPreferences(PRIVACY_PREFS, Context.MODE_PRIVATE)
            return PrivacySettings(prefs)
        }
    }
}

File: apps/mobile/shared/Sources/FlagshipCore/PrivacySettings.swift
import Foundation
import Observation

/// B12 — persisted user preferences governing how the app gates
/// access to its content. Today the only toggle is biometric-at-launch;
/// future additions (auto-lock-after-N-minutes, lock-on-screenshot,
/// etc.) live alongside.
///
/// The gate state itself lives in AppState (the in-memory latch),
/// not here — this object only owns the PERSISTED user choice.
/// AppState reads this on launch to decide the initial latch value.
@Observable
@MainActor
public final class PrivacySettings {
    /// App appearance: follow the system, or force light / dark. Stored as a
    /// raw string; the SwiftUI `ColorScheme?` mapping lives in the UI layer
    /// (this module is Foundation-only). `auto` ⇒ no override (system-derived).
    public enum ThemeMode: String, CaseIterable, Sendable {
        case auto
        case light
        case dark
    }

    private let defaults: UserDefaults
    private let requireBiometricKey = "flagship.privacy.requireBiometric"
    private let requirePassphraseKey = "flagship.privacy.requirePassphrase"
    private let themeModeKey = "flagship.appearance.themeMode"

    /// True if a Face ID / Touch ID evaluation is required each time the
    /// app cold-launches or returns from background. Defaults to TRUE —
    /// a restored account opens behind a face-unlock rather than a full
    /// sign-in. Only gates once an account is paired (a first-ever launch
    /// with no account shows Welcome unguarded), and the device's own
    /// "Sign out" fallback covers a phone with no biometric enrolled. The
    /// user can turn this off in Settings → Privacy to open straight in.
    public var requireBiometricAtLaunch: Bool {
        didSet { defaults.set(requireBiometricAtLaunch, forKey: requireBiometricKey) }
    }

    /// The strictest option: when ON, the app does NOT restore the
    /// persisted session on launch, so every open requires a full
    /// sign-in (the account passphrase), not just a Face ID unlock.
    /// Default OFF — most users want to stay signed in behind Face ID.
    /// Supersedes `requireBiometricAtLaunch` when both are set (a full
    /// sign-in is strictly stronger than a face-unlock).
    public var requirePassphraseAtLaunch: Bool {
        didSet { defaults.set(requirePassphraseAtLaunch, forKey: requirePassphraseKey) }
    }

    /// Chosen app appearance. Default `auto` (follow the system). Persisted on
    /// every change; the UI layer applies it via `.preferredColorScheme`.
    public var themeMode: ThemeMode {
        didSet { defaults.set(themeMode.rawValue, forKey: themeModeKey) }
    }

    public init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        // Unset → default ON. Once the user makes an explicit choice we
        // honour the stored value.
        if defaults.object(forKey: requireBiometricKey) == nil {
            self.requireBiometricAtLaunch = true
        } else {
            self.requireBiometricAtLaunch = defaults.bool(forKey: requireBiometricKey)
        }
        // Unset → default OFF (opt-in to the slower, stricter mode).
        self.requirePassphraseAtLaunch = defaults.bool(forKey: requirePassphraseKey)
        // Unset → default `auto` (follow the system appearance).
        self.themeMode = ThemeMode(rawValue: defaults.string(forKey: themeModeKey) ?? "") ?? .auto
    }
}
